start.sh

```diff
@@ -38,25 +76,25 @@ fi
 
 start_ethminer
 sleep 2
-# Make sure DAG generation completes before starting livepeer
-echo "Waiting for DAG generation..."
-dag_generated=0
-while [ $dag_generated == 0 ]
+
+dag_generation_check
+
+start_livepeer
+
+## Make sure we stop and start livepeer during any future DAG generation, which happens at an epoch_change
+epoch_changes_old=$(echo '{"id":0,"jsonrpc":"2.0","method":"miner_getstatdetail"}' | netcat -w 1 127.0.0.1 3333 | jq .result.mining.epoch_changes)
+while true
 do
-    # Assumptions:
-    # 1. If hashrate is greater than 0 then DAG generation is complete
-    # 2. The DAG will be loaded on the GPUs in parallel (default ethminer behavior) 
-    # Note: If the 2nd assumption is not true then it is possible for the overall hashrate to be greater than 0, but the DAG
-    # not to be loaded on all GPUs yet
-    hashrate=$(echo '{"id":0,"jsonrpc":"2.0","method":"miner_getstatdetail"}' | netcat -w 1 127.0.0.1 3333 | jq .result.mining.hashrate)
-    if [ $hashrate != "\"0x00000000\"" ]
+    epoch_changes_new=$(echo '{"id":0,"jsonrpc":"2.0","method":"miner_getstatdetail"}' | netcat -w 1 127.0.0.1 3333 | jq .result.mining.epoch_changes)
+    if [[ $epoch_changes_new -gt $epoch_changes_old ]]
     then
-        dag_generated=1
+        echo "New Epoch detected, stopping livepeer for DAG generation"
+        stop_livepeer
+        epoch_changes_old=$epoch_changes_new
+        sleep 10
```

Comment:
> Currently the sleep time for each epoch_changes poll is set to 1 second, and once we detect an epoch change, there is a sleep for 10 seconds before we start polling again to check if DAG generation is complete.
> I am not sure if these values are OK for prod though. Should these be configurable through an environment variable?

In response to the above comment:

Yeah just to be safe and for flexibility I think it could make sense to use an environment variable `DAG_GENERATION_TIME` so the script user can adjust it if needed. To be absolutely certain that an edge case where the script sleeps and then detects a non-zero hashrate while DAG generation is still running can be avoided, `DAG_GENERATION_TIME` could just be set to the estimated DAG generation time which can be determined during benchmarking. In the case of a 1660, we've observed DAG generation time to be around 35s.